### PR TITLE
Library to handle Merkle trees on-chain

### PR DESCRIPTION
**Pull request by NouDaimon**
Contains:
- `MerkleTree` struct
- `updateNode` function to update a leaf node hash and the associated parent hashes 

Potential changes:
- Make `updateNode` private
- Add `remove`, `update` and `add` functions, which use `updateNode`

### CODE REVIEW

**Review by ItsNickBarry — COMMENTED** [review]
@NouDaimon Added an alternative implementation.  It's similar to yours, but I think it's simpler.  Also added some comments.

File: contracts/data/IncrementalMerkleTree2.sol [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+pragma solidity ^0.8.0;
+
+import 'hardhat/console.sol';
+
+library IncrementalMerkleTree2 {
+    using IncrementalMerkleTree2 for Tree;
+
+    struct Tree {
+        bytes32[][] nodes;
```
Comment (ItsNickBarry) [inline contracts/data/IncrementalMerkleTree2.sol:11]: @NouDaimon I think the main disadvantage of this approach is the lack of support an `indexOf` function.  However, I don't think it's a fundamental part of the data structure, and end users could add their own implementations if they want to spend the gas.

**Comment by NouDaimon** [comment]
Several implementations of Merkle trees always mantain a balanced tree - from what I have seen, this is either done via initialization or duplication. 

 An example is the `Progressive Merkle Tree` for the new ETH deposit contract  (denoted in the image [here](https://www.researchgate.net/figure/Illustration-of-the-incremental-Merkle-tree-algorithm-Node-numbers-are-labeled-in-the_fig1_342969698))which initializes all leaves as `hash(0)` and calculates the root based on these hashes, as is normally done in Merkle Trees.
 
 Another example is the `Bitcoin` implementation (denoted[ here](https://medium.com/coinmonks/merkle-trees-concepts-and-use-cases-5da873702318)) which duplicates the leaf node, should a pair not exist, and hashes upwards to calculate the root as normal.
 
 The final example for `Monero` (again denoted [here](https://medium.com/coinmonks/merkle-trees-concepts-and-use-cases-5da873702318)) seems to follow the same implementation as the one in the SolidState `IncrementalMerkleTree.sol` library; the appended node is bubbled upwards to calculate a new root from itself and the old root. 
 
 Are there any security considerations for not initializing the `MerkleTree` or growing it in this fashion? Any considerations/potential attack vectors from using a 'dynamic height' and unbalance tree?

File: test/data/IncrementalMerkleTree.ts [hunk 2]
```diff
@@ -121,6 +121,24 @@ describe('IncrementalMerkleTree', function () {
 
       expect(await instance.callStatic.root()).to.equal(hash);
     });
+
+    it('POC: returns Merkle root derived from elements contained in unbalanced tree', async () => {
```
Comment (ItsNickBarry) [inline test/data/IncrementalMerkleTree.ts:125]: What is "POC"?

Comment (ItsNickBarry) [reply test/data/IncrementalMerkleTree.ts:125]: This test seems a bit redundant.  Though, there's a lot of overlap in several of the tests.

**Comment by ItsNickBarry** [comment]
Need confirmation that the improved hashing function will not cause issues for end-users who might be working with data in memory.

Otherwise, I think this is complete.

**Comment by NouDaimon** [comment]
> Need confirmation that the improved hashing function will not cause issues for end-users who might be working with data in memory.
> 
> Otherwise, I think this is complete.

As per the discussion, given the [docs](https://docs.soliditylang.org/en/v0.8.17/internals/layout_in_memory.html?highlight=free%20memory%20slot#layout-in-memory), the memory space being used in the hash function is "scratch space" so should not affect end users memory.

Auditor should confirm.

**Comment by ItsNickBarry** [comment]
I updated so that the tree accepts hashes directly, rather than raw data.  The end user must do the hashing.

**Review by hinterlist — APPROVED** [review]

**Comment by ItsNickBarry** [comment]
@NouDaimon I'm merging this now.  We can merge the optimizations later, after further review or audit.

Thanks @hinterlist.

**Pull request merged**